fix(LSM): restructure Compaction to cleanup and handle concurrent IO / CPU asserts

kprotty opened the pull request:
Restructured the `Compaction` struct to clean up control flow and get rid of various asserts being tripped regarding concurrent IO and CPU work. This is the first step to addressing the `table_count_visible` assertion hit in VOPR seed 545.

Commit:
LSM: Fix assertion in Manifest.remove_invisible_tables + reuse TableInfoBuffer

The assertions were overly tight and assumed the tables reported from the iterator intersected with the key range hint.
Instead, the tables should only overlap as asserted elsewhere.

The function also implemented its own form of TableInfoBuffer() so went ahead and used the shared abstraction.

jorangreef commented:
@kprotty what is the next step to get this fixed? What can we try here?

Commit:
LSM: Restructure Compaction

Tried to restructure the Compaction type implementation as an attempt to simplify control flow for analysis.
This currently errors out with a rogue Grid.Write having its callback triggered.

Commit:
LSM: fix Compaction grid writes + cleanup Tree compact code

I was using `inline for` to generate the callbacks for the Compaction's BlockWrites to Grid. This had unexpected behavior where it cached the callback for the first field so the other fields used the wrong one. Moving the per-field writing to its own function solved this.

Also took the opportunity to clean up the control flow and assert orderings in the Tree compact code:
- Now, invisible tables are removed from `level_b` and happen when a compaction completes instead of when all compactions for that half measure complete.
- The `mut_table.can_commit_batch` assert was also moved after mut_table is sorted into immut_table.

kprotty commented:
@jorangreef I've pivoted this PR into restructuring the Compaction and addressing various asserts that were hit along the way.

Commit:
LSM: Fix Compaction BlockWrite/update_tables assumptions

Given IO is now serialized with CPU merge, we don't need the intermediary blocks for index, data, and filter. They now point directly do the Table.Builder's blocks for writing.

We also know that tables reported by the iterator_b (LevelIterator) will always be visible to the compaction snapshot so they can be queued for removal/updating.

Finally, moved around the check for a full data block to after builder.data_block_finish() was called as that's when it writes the block's vsr.Header to work correctly with the table helpers.

Commit:
LSM: Remove exclusive snapshot range for table visibility

Since snapshots are no longer as unique as they once were (with compaction updating with snapshots from the past), the visibility asserts don't have to be as strict in their ranges.

Commit:
LSM: Compaction release grid blocks for tables in iterator_b

These tables will be deleted so their Grid blocks can be re-used. It should also be safe to release their blocks in the iterator_b callback as iterator_b (LevelIterator) makes a copy of the block contents itself.

Commit:
LSM: Compaction remove level_a table once merged

The level_a selected table and the overlapping level_b tables are merged into new tables inserted into level_b. The old input level_b tables were already being removed, so this removes the selected level_a one as well.

Commit:
LSM: use ManifestLog use grid to acquire/release blocks from free_set

The manifest_log was using the superblock's free_set directly to acquire/release block addresses. This conflicts with it also using the Grid for read/write.

Grid read/writes require that the block addresses you pass in to be from its own acquire/release since those update its cache. The ManifestLog was not doing this so we address that here.

Commit:
LSM: Ignore empty Grid read iops when iterating.

Grid read iops can be observed to have empty `read` FIFOs during Grid.read_block_callback. There, the if branch pops from the FIFO and invokes their callbacks. On the last pop, the FIFO can be empty and the callback could iterate the iops expecting their FIFO's to not be empty.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -26,70 +27,97 @@ pub fn CompactionType(
+            /// started are included in the compaction. If this is not the case, the older table
+            /// in a pair of overlapping tables could be left in level A and shadow the newer table
+            /// in level B, resulting in data loss/invalid data.
+            fn precedence(compaction: *Compaction, stream_a: u32, stream_b: u32) bool {
```
```suggestion
            fn precedence(compaction: *const Compaction, stream_a: u32, stream_b: u32) bool {
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -26,70 +27,97 @@ pub fn CompactionType(
+            /// This assumes that all overlapping tables in level A at the time the compaction was
+            /// started are included in the compaction. If this is not the case, the older table
+            /// in a pair of overlapping tables could be left in level A and shadow the newer table
+            /// in level B, resulting in data loss/invalid data.
```
I don't understand what this is saying; can you clarify it a little? Right now it reads like level A's tables might not have disjoint key ranges (but I think they do). Unless we are talking about snapshots?

kprotty replied:
This comment was copy-pasted from the original version. I can clarify

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -126,125 +145,94 @@ pub fn CompactionType(
-            manifest: *Manifest,
-            // TODO level_a_table: ?TableInfo,
-            level_b: u8,
+            table: ?*const TableInfo,
```
```suggestion
            table_a: ?*const TableInfo,
```
Just to be explicit that this corresponds to `iterator_a_context`.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -126,125 +145,94 @@ pub fn CompactionType(
+            compaction.iterator_a.deinit(allocator);
         }
 
         pub fn start(
```
```suggestion
        /// `table_a` is null when the level A is 0.
        pub fn start(
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -126,125 +145,94 @@ pub fn CompactionType(
             assert(compaction.status == .idle);
             assert(compaction.callback == null);
             assert(compaction.io_pending == 0);
+
```
```suggestion
assert((table == null) == (level_b - 1 == 0));
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -126,125 +145,94 @@ pub fn CompactionType(
             assert(range.table_count > 0);
 
+            // Any level may choose to drop tombstones, but the last level must do so 
+            // as it would have nowhere else to move/compact the tables to.
```
> Any level may choose to drop tombstones

Levels must only drop tombstones for keys which no lower levels include. Otherwise, dropping a tombstone may "reveal" an old value.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -126,125 +145,94 @@ pub fn CompactionType(
-
-            // TODO Reset builder.
-
-            // TODO: Enable when move_table() can fetch TableInfo from address/checksum.
-            //
-            // Perform a "compaction move" to the next level inline if certain factors allow:
-            // - Can only do the specialization if there's a single table to compact.
-            // - Must be compacting from a table iterator which has an address and checksum.
-            // - Cannot drop tombstones as then we have to go through the normal compaction path.
-            // - Cannot be performing the immutable table -> level 0 compaction
-            //   as it requires the table being moved to reside on disk (tracked by manifest).
-            if (false and IteratorA.Context == TableIteratorType(Table, Storage)) {
-                if (!drop_tombstones and range.table_count == 1) {
-                    assert(compaction.level_b != 0);
-                    assert(compaction.status == .compacting);
-
-                    const level_a = level_b - 1;
-                    assert(level_a < config.lsm_levels - 1);
-
-                    compaction.manifest.move_table(
-                        level_a,
-                        level_b,
-                        snapshot,
-                        iterator_a_context.address,
-                        iterator_a_context.checksum,
-                    );
-
-                    compaction.status = .done;
-                    return;
-                }
-            }
```
Is this "compaction move" not going to be supported? (I realize it wasn't fully implemented).

kprotty replied:
It was an optimization that relied on the manifest being able to lookup the table using the address + checksum which we never got to implementing.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -254,11 +242,21 @@ pub fn CompactionType(
+            compaction.iterator_b.start(iterator_b_context, iterator_b_io_callback);
+        }
+
+        fn iterator_a_io_callback(iterator_a: *IteratorA) void {
```
I like the `_io_callback` name! Easier to follow, and it nicely disambiguates it from the other table info callback.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -26,70 +27,97 @@ pub fn CompactionType(
         snapshot: u64,
         drop_tombstones: bool,
 
+        status: Status,
```
Can you add documentation explaining `status` and `merge_status` (and the difference/relationship between them).

kprotty replied:
* `status` is used to know when the compaction is started/done over many ticks.
* `merge_status` is used to know when to initialize MergeIterator and replaces the previous `merge_done` flag. I think this can be removed/simplified.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -267,41 +265,55 @@ pub fn CompactionType(
-                }
-            }
+            // Release the table's block addresses in the Grid as it will be made invisible.
+            // This is safe to do so as iterator_b makes a copy of the block before calling us.
```
```suggestion
            // This is safe; iterator_b makes a copy of the block before calling us.
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -267,41 +265,55 @@ pub fn CompactionType(
             const tables: []TableInfo = buffer.drain();
             if (tables.len == 0) return;
 
+            // Double check that the tables queued are in the compaction's key range.
```
```suggestion
            // Double check that the queued tables are in the compaction's key range.
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -311,293 +323,256 @@ pub fn CompactionType(
-            if (compaction.ticks >= pipeline_tick_merge) {
-                if (compaction.merge_iterator.empty()) {
-                    assert(!compaction.merge_done);
+        fn io_write_start(compaction: *Compaction, comptime field: []const u8) void {
```
To make this a little more self-documenting, maybe use a `enum { data, filter, index }` for `field`'s type (and then use `@field(compaction, @tagName(field))` down below).

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -26,70 +27,97 @@ pub fn CompactionType(
-        /// because a write I/O may yet follow even after the merge is done.
-        merge_done: bool = false,
+        merge_status: Status,
         merge_iterator: MergeIterator,
```
Make this a `?MergeIterator` so that we can assert exactly when it does/doesn't exist.

e.g. just before
```
compaction.merge_iterator = MergeIterator.init(compaction, k, .ascending);
```
we could `assert(compaction.merge_iterator == null);`.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -126,125 +145,94 @@ pub fn CompactionType(
-                .merge_iterator = undefined, // This can only be initialized on tick 1.
-                .table_builder = table_builder,
+                .merge_status = .idle,
+                .merge_iterator = undefined, // Assigned by start()
```
```suggestion
                .merge_iterator = undefined, // Assigned by cpu_merge_start()
```

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -26,70 +27,97 @@ pub fn CompactionType(
+        const BlockWrite = struct {
+            write: Grid.Write = undefined,
+            block: BlockPtr = undefined,
+            ready: bool = false,
```
nit: What do you think of renaming `ready` to `done`? `ready` could mean "ready for more data" or "ready to be written". Or it could be "closed" (i.e. a "closed" block can't receive any more values, and open block can receive values).

kprotty replied:
A block is `ready` when it can be written out on the next `compact_tick`. Blocks are made `ready` by `cpu_merge()`

sentientwaffle replied:
I understand; I was just suggesting a less ambiguous name for the field.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -311,293 +323,256 @@ pub fn CompactionType(
+            // Double check that manifest updates have been applied.
+            assert(compaction.update_level_b.drain().len == 0);
+            assert(compaction.insert_level_b.drain().len == 0);
 
```
Can we also re-assert that the iterators are all done?

kprotty replied:
This happens above

sentientwaffle commented on src/lsm/grid.zig:
```diff
@@ -207,7 +207,8 @@ pub fn GridType(comptime Storage: type) type {
                 var it = grid.read_iops.iterate();
                 while (it.next()) |iop| {
-                    assert(address != iop.reads.peek().?.address);
+                    const iop_read = iop.reads.peek() orelse continue;
```
Why did this change? I thought there was an invariant that a `read_iop` always has at least one `read` in its queue (the read that triggered the iop).

kprotty replied:
I put the reasoning in the commits without as well being in the source (my mistake). This invariant is true up until [the last loop where the IOP's `read` FIFO is being dequeued](https://github.com/tigerbeetledb/tigerbeetle/blob/main/src/lsm/grid.zig#L417-L421). This last dequeue calls the read's callback which could enter one of the grid functions and iterate the IOPs again (expecting non-empty `read` FIFOs). I decided to forgo that invariant as the imagined control flow for handling the `release()` of the empty IOP before that last callback would've been awkward.

sentientwaffle commented on src/lsm/manifest.zig:
```diff
@@ -342,39 +346,37 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
             assert(level < config.lsm_levels);
             assert(compare_keys(key_min, key_max) != .gt);
 
+            const direction = .ascending;
```
Doesn't this need to be `descending` to prevent the iterator getting out of sync when tables are removed?

kprotty replied:
Yea, forgot about that discussion. Although I believe that the tables recorded into the buffer will need to be reversed before calling `remove_tables()` as that expects them to be ascending

sentientwaffle replied:
Good point! Yes, it does need to be reversed.

sentientwaffle commented on src/lsm/tree.zig:
```diff
@@ -618,37 +614,33 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
 
-            // Try to tick the cpu portion of the immutable table compaction:
+            // Always start one io_pending that is resolved right after
+            // to handle the case of no level or immutable table being ticked for compaction
```
```suggestion
            // Always start one fake io_pending that is resolved right after
            // to handle the case where this compaction tick triggers no IO.
            // (For example, ticking the immutable table, or level B is already done).
```
```

sentientwaffle commented on src/lsm/tree.zig:
```diff
@@ -618,37 +614,33 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
             );
         }
 
-        fn compact_io(tree: *Tree) void {
```
The docs in this file still make many references to `compact_io`, can you update them accordingly?

eatonphil commented on src/lsm/manifest.zig:
```diff
@@ -50,9 +54,9 @@ pub fn TableInfoType(comptime Table: type) type {
-            assert(snapshot != table.snapshot_max);
+                
+            // assert(snapshot != table.snapshot_min);
+            // assert(snapshot != table.snapshot_max);
```
Are these comments intentional/permanent?

kprotty replied:
Intentional and explained in commit. I'll add the explanation to source code as well.

Commit:
LSM: use const pointers for Merge stream callbacks

Commit:
LSM: clarify table_a compaction usage + remove unused move_table()

Commit:
LSM: more Compaction table_a fixes + cleanup merge_status

Brings back merge_done and makes MergeIterator optional to represent .idle state.

Commit:
LSM: Rename Compaction.BlockWrite.ready to writable

Commit:
LSM: replace compact_io() refererences with compact_tick()

Commit:
LSM: Remove invisible tables in reverse to avoid iterator invalidation.

While scanning for invisible tables in a level, we may fill up the table buffer and flush/remove them from the level. When scanning in ascending order, this could invalidate the references from the iterator.

Instead, scan in descending order to ensure the tables we queue and potentially flush/remove have already been visited.

Commit:
LSM: Update doc comments & clarify snapshot visibility

There's a slight change here in that snapshot_min check is now inclusive while snapshot_max remains exclusive.

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -311,293 +323,257 @@ pub fn CompactionType(
-                assert(tombstones_dropped > 0);
-            } else {
+            // Finalize the data block and prepare it to be written out.
+            blk: {
```
What about:

```zig
if (compaction.table_builder.data_block_full() or
    compaction.table_builder.filter_block_full() or
    compaction.table_builder.index_block_full() or
    compaction.merge_iterator.?.empty())
{
    if (compaction.table_builder.data_block_empty()) {
        ... break :blk;
    }
```

— this way if `drop_tombstones` is `true` and iterator A has many tombstones, we won't create a bunch of tiny data blocks from values that _aren't_ dropped. Instead we'll collect them between ticks.

sentientwaffle commented on src/lsm/tree.zig:
```diff
@@ -618,37 +614,34 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
 
-        fn compact_io_tick(tree: *Tree, compaction: anytype) void {
+        fn compact_tick(tree: *Tree, compaction: anytype) void {
+            if (compaction.status != .processing) return;
```
If the compaction's status isn't `processing`, can we assert that it is `done`?

kprotty replied:
Here, it could be `idle` if it was never started, or `done` if has fully completed. This function ticks a compaction if it's still `processing`. That status check can be moved to the caller though.

sentientwaffle replied:
I think it is ok where it is, I just didn't realize that a compaction could be `idle` here.

sentientwaffle commented on src/lsm/tree.zig:
```diff
@@ -693,155 +687,112 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
-                } else if (tree.compaction_table_immutable.status == .compacting) {
-                    still_compacting = true;
+                switch (tree.compaction_table_immutable.status) {
+                    .idle => {}, // wasn't started for this half measure
```
If we didn't start it this half measure, I assume the immutable table is empty? Can we assert `tree.compaction_table_immutable.free`?

Commit:
LSM: Gather more merged values before writing blocks.

The old pattern would enforce that pending blocks at a tick are written out on the next tick. This instead gathers pending blocks across multiple ticks until they either get full or there's no more values to merge.

Commit:
LSM: assert immuatble table is free when its compaction is idle

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -311,293 +323,234 @@ pub fn CompactionType(
+            // when there's no more merged values to fill them.
+            if (compaction.table_builder.index_block_full() or
+                (merge_iterator.empty() and !compaction.table_builder.index_block_empty())
+            ) {
```
:+1: Perfect.

Fwiw I think our formatting convention is:
```suggestion
            if (compaction.table_builder.index_block_full() or
                (merge_iterator.empty() and !compaction.table_builder.index_block_empty()))
            {
```

Commit:
LSM: Tighten table info buffer sizes when compacting.

They're allocated with the expected maximum so overflow to flush to manifest mid-enqueue should never occur.

Commit:
LSM: Fix Compaction cpu_merge() if block formatting

sentientwaffle commented on src/lsm/compaction.zig:
```diff
@@ -126,125 +140,101 @@ pub fn CompactionType(
+            assert(!compaction.merge_done and compaction.merge_iterator == null);
+
             assert(range.table_count > 0);
+            assert(range.table_count <= compaction.update_level_b.array.len);
```
Doesn't this need a `+ 1` on the right? I think `CompactionRange` includes the level A table.

kprotty replied:
Good catch. What about ` + @boolToInt(table_a != null)`?

sentientwaffle replied:
I think even when `table_a` is null (i.e. we are compacting the immutable table), from the `CompactionRange`'s point of view it is still counted (see: https://github.com/tigerbeetledb/tigerbeetle/blob/dfb67bd95a528bc9a47ba16ccb6111c39c212707/src/lsm/tree.zig#L551-L557).

Commit:
LSM: Fix range.table_count assertion in Compaction
